From 2c7e169aee241b0dc252f6ba5283f7e8691e0e51 Mon Sep 17 00:00:00 2001
From: Wang Wensheng <wangwensheng4@huawei.com>
Date: Thu, 4 May 2023 07:16:35 +0000
Subject: [PATCH 41/53] [Huawei] mm/sharepool: Take apart mm_add_group_init

Offering: HULK
hulk inclusion
category: feature
bugzilla: N/A

--------------------------------

There are two branches in mm_add_group_init() that either to initialize
the sp_group_master for the new task or to check if the new task is
proper to be added into the group.
The first piece is in the categroy of initialization for new task and
has nothing to do with the sp_group. The second part is a joint
inspection together the sp_group and the new task. Take them apart to
make the code more reasonable.

Signed-off-by: Wang Wensheng <wangwensheng4@huawei.com>
---
 mm/share_pool.c | 85 +++++++++++++++++++++++++------------------------
 1 file changed, 44 insertions(+), 41 deletions(-)

diff --git a/mm/share_pool.c b/mm/share_pool.c
index e940345bc9a3..b8e82bb1f962 100644
--- a/mm/share_pool.c
+++ b/mm/share_pool.c
@@ -515,8 +515,8 @@ static int sp_group_setup_mapping(struct mm_struct *mm, struct sp_group *spg)
 
 static struct sp_group *sp_group_create(int spg_id);
 static void free_sp_group_locked(struct sp_group *spg);
-static int group_add_task(struct mm_struct *mm, struct sp_group *spg,
-					    unsigned long prot);
+static int sp_group_link_task(struct mm_struct *mm, struct sp_group *spg,
+			      unsigned long prot);
 static int init_local_group(struct mm_struct *mm)
 {
 	int ret;
@@ -528,7 +528,7 @@ static int init_local_group(struct mm_struct *mm)
 		return PTR_ERR(spg);
 
 	master->local = spg;
-	ret = group_add_task(mm, spg, PROT_READ | PROT_WRITE);
+	ret = sp_group_link_task(mm, spg, PROT_READ | PROT_WRITE);
 	if (ret)
 		/* The spm would be released while destroying the spg */
 		goto free_spg;
@@ -1191,33 +1191,6 @@ static struct sp_group *find_or_alloc_sp_group(int spg_id)
 	return spg;
 }
 
-/* the caller must hold sp_global_sem */
-static int mm_add_group_init(struct task_struct *tsk, struct mm_struct *mm,
-			     struct sp_group *spg)
-{
-	int ret;
-	struct sp_group_master *master;
-
-	if (!mm->sp_group_master) {
-		ret = sp_init_group_master_locked(tsk, mm);
-		if (ret)
-			return ret;
-	} else {
-		if (is_process_in_group(spg, mm)) {
-			pr_err_ratelimited("task already in target group, id=%d\n", spg->id);
-			return -EEXIST;
-		}
-
-		master = mm->sp_group_master;
-		if (master->group_num == MAX_GROUP_FOR_TASK) {
-			pr_err("task reaches max group num\n");
-			return -ENOSPC;
-		}
-	}
-
-	return 0;
-}
-
 /* the caller must hold sp_global_sem */
 static struct sp_group_node *create_spg_node(struct mm_struct *mm,
 	unsigned long prot, struct sp_group *spg)
@@ -1275,12 +1248,32 @@ static void free_spg_node(struct mm_struct *mm, struct sp_group *spg,
 	kfree(spg_node);
 }
 
-/* the caller must hold sp_global_sem and down_write(&spg->rw_lock) in order */
-static int group_add_task(struct mm_struct *mm, struct sp_group *spg,
-			  unsigned long prot)
+/*
+ * sp_group_link_task - Actually add a task into a group
+ * @mm: specify the input task
+ * @spg: the sp_group
+ * @prot: read/write protection for the task in the group
+ *
+ * The input @mm and @spg must have been initialized properly and could not
+ * be freed during the sp_group_link_task().
+ * the caller must hold sp_global_sem and down_write(&spg->rw_lock) in order.
+ */
+static int sp_group_link_task(struct mm_struct *mm, struct sp_group *spg,
+			      unsigned long prot)
 {
-	struct sp_group_node *node;
 	int ret;
+	struct sp_group_node *node;
+	struct sp_group_master *master = mm->sp_group_master;
+
+	if (is_process_in_group(spg, mm)) {
+		pr_err("task already in target group, id=%d\n", spg->id);
+		return -EEXIST;
+	}
+
+	if (master->group_num == MAX_GROUP_FOR_TASK) {
+		pr_err("task reaches max group num\n");
+		return -ENOSPC;
+	}
 
 	ret = sp_group_setup_mapping(mm, spg);
 	if (ret)
@@ -1317,6 +1310,13 @@ static int group_add_task(struct mm_struct *mm, struct sp_group *spg,
  * SPG_ID_AUTO:
  * 		the task would be added into a new group with a new id in range
  *		[SPG_ID_AUTO_MIN, SPG_ID_AUTO_MAX].
+ *
+ * This function can be taken into four parts:
+ * 1. Check and initilize the task specified by @tgid properly.
+ * 2. Create or get the spg specified by @spg_id.
+ * 3. Check the spg and task together and link the task into the spg if
+ *    everything looks good.
+ * 4. Map the existing sp_area from the spg into the new task.
  */
 int mg_sp_group_add_task(int tgid, unsigned long prot, int spg_id)
 {
@@ -1376,6 +1376,15 @@ int mg_sp_group_add_task(int tgid, unsigned long prot, int spg_id)
 		goto out_put_task;
 	}
 
+	if (!mm->sp_group_master) {
+		/* We don't free mm->sp_group_master on failure */
+		ret = sp_init_group_master_locked(tsk, mm);
+		if (ret) {
+			up_write(&sp_global_sem);
+			goto out_put_mm;
+		}
+	}
+
 	if (mm->sp_group_master && mm->sp_group_master->tgid != tgid) {
 		up_write(&sp_global_sem);
 		pr_err("add: task(%d) is a vfork child of the original task(%d)\n",
@@ -1392,13 +1401,7 @@ int mg_sp_group_add_task(int tgid, unsigned long prot, int spg_id)
 	}
 
 	down_write(&spg->rw_lock);
-	ret = mm_add_group_init(tsk, mm, spg);
-	if (ret) {
-		up_write(&spg->rw_lock);
-		goto out_drop_group;
-	}
-
-	ret = group_add_task(mm, spg, prot);
+	ret = sp_group_link_task(mm, spg, prot);
 	if (ret) {
 		up_write(&spg->rw_lock);
 		goto out_drop_group;
-- 
2.17.1

